New contexts for `StoreNoticesContainer` and notice grouping

mikejolley opened the pull request:
This PR refactors `StoreNoticeContainer` so it combines snack bars and regular notices (instead of separate components) and introduces new `noticeContexts` for inner blocks. 

**New noticeContexts**

Adds several notice contexts for the main inner blocks that can be used as `context` instead of strings. Each notice context has a corresponding `StoreNoticesContainer` component to render any notices added.

Rather than create a new interface, I extended the existing one which only contained payment contexts:

```js
PAYMENTS = 'wc/checkout/payments',
EXPRESS_PAYMENTS = 'wc/checkout/express-payments',
```

The full list is now:

```js
CART = 'wc/cart',
CHECKOUT = 'wc/checkout',
PAYMENTS = 'wc/checkout/payments',
EXPRESS_PAYMENTS = 'wc/checkout/express-payments',
CONTACT_INFORMATION = 'wc/checkout/contact-information',
SHIPPING_ADDRESS = 'wc/checkout/shipping-address',
BILLING_ADDRESS = 'wc/checkout/billing-address',
SHIPPING_METHODS = 'wc/checkout/shipping-methods',
CHECKOUT_ACTIONS = 'wc/checkout/checkout-actions',
```

**Notice grouping**

Groups notices of the same type into a single error notice. You can test this on the cart page by calling this a few times:

```
wp.data.dispatch( 'core/notices' ).createNotice( 'error', 'Test', { context: 'wc/cart' } );
```

![Screenshot 2022-12-14 at 16 44 13](https://user-images.githubusercontent.com/90977/207655928-9747f4db-e4c7-4603-8366-5109cf562886.png)

**Notice creation helpers**

[Existing notices have been modified](https://github.com/woocommerce/woocommerce-blocks/pull/7728) to use the new containers and `createNotice` functions.

Additionally, when a new error is displayed, it will scroll to the parent `StoreNoticeContainer`. The scrolling fixes #2559

This mechanism also ensures that if a `StoreNoticesContainer` does not exist, it uses a parent context.

**Changes to hooks**

**New filters: `showApplyCouponNotice` and `showRemoveCouponNotice`**

To simplify notice handling, I've moved the error suppression for coupons to `showApplyCouponNotice` to prevent the notice from being added in the first place. This will break Points and Rewards (in a non-destructive way), so this will need a patch to use the new filter (https://github.com/woocommerce/woocommerce-points-and-rewards/blob/5d3b6812188e969505f612868ce228a387f1ea12/assets/js/filters.js#L24).

The fix for points and rewards can be found here: https://github.com/woocommerce/woocommerce-points-and-rewards/pull/491

**Other changes**

- Removed `useCheckoutNotices`. It was only used in one place and not shared, so I moved the logic inline.
- `removeAllNotices` when processing checkout so new notices can surface.
- Typed checkout events and push-changes code.
- Helpers to `processErrorResponse` and `processInvalidParamResponse` from the API

Closes #2559

### Testing

### User-facing testing

Snack bars:

1. Add a product to the cart.
2. Go to cart page
3. Apply a valid coupon code
4. See snack bar notice on the page

Error notices from API:

1. Add a product to the cart.
5. Go to the cart page and set the quantity to 5 for the product.
6. Checkout. Fill out the form but do not submit.
7. In another tab, go into product admin and edit the product. Enable stock management, and set stock qty to 4.
8. Go back to the checkout and place order.
9. Checkout should scroll to the top and display an error:

![Screenshot 2022-12-14 at 16 22 28](https://user-images.githubusercontent.com/90977/207651044-ed38cf35-f9b5-4eb9-ba69-79bfce7a1b65.png)

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (i.e.: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

cc @opr

mikejolley committed:
Refactor Store Notices

Move snackbar hiding filter before notice creation

Implements showApplyCouponNotice

Refactor context providers

Use STORE_NOTICE_CONTEXTS

use refs to track notice containers

Refactor ref usage

Use existing noticeContexts

mikejolley committed:
Move new notice code to checkout package

mikejolley committed:
Combine store and snackbars

mikejolley committed:
Update noticeContexts imports

mikejolley committed:
Remove context provider

mikejolley committed:
Update data store

mikejolley committed:
Fix 502

mikejolley committed:
Add new error contexts

mikejolley committed:
Force types

mikejolley committed:
Unnecessary reorder of imports

mikejolley committed:
Fix global handling

opr commented on packages/checkout/components/store-notices-container/snackbar-notices.tsx:
```diff
@@ -0,0 +1,53 @@
+
+const SnackbarNotices = ( {
+	className,
+	context,
```
JS linting error here: `[@typescript-eslint/no-unused-vars] 'context' is defined but never used.`

opr commented on packages/checkout/components/store-notices-container/types.ts:
```diff
@@ -0,0 +1,17 @@
+	context?: string;
+	showGlobal: boolean;
+	additionalNotices?: ( NoticeType & NoticeOptions )[];
+	forceType: 'default' | 'snackbar' | null;
```
Can we add a comment to document this please? I understand it's being used to force the notices shown in this component to be a certain type (so if something _was_ a regular notice, but `forceType` is `snackbar` it'll instead turn into a snackbar notice, is this correct?

opr commented on packages/checkout/components/store-notices-container/types.ts:
```diff
@@ -0,0 +1,17 @@
+
+export interface StoreNoticesContainerProps {
+	className?: string;
+	context?: string;
```
If this is an optional property, you should also specify `| undefined` as its type because of this rule https://github.com/woocommerce/woocommerce-blocks/blob/d9f0aa8b8e26bbf50f109ac1f845b22ac8be208b/tsconfig.base.json#L10

same for the other optional props

opr commented on packages/checkout/utils/create-notice.ts:
```diff
@@ -0,0 +1,58 @@
+	const registeredContext = Object.keys( containerRefs );
+
+	// If the container ref was not registered, use the parent context instead.
+	if ( ! registeredContext.includes( noticeContext ) ) {
```
What do you think about abstracting this to a function to avoid these if elses?

opr commented on packages/checkout/components/store-notices-container/store-notices.tsx:
```diff
@@ -0,0 +1,85 @@
+	// Register the container with the parent.
+	useEffect( () => {
+		registerContainer( context, ref );
+		return function cleanup() {
```
how come you've named this function? for readability or is there another reason?

mikejolley replied:
Yeah I just did it based on an example I found. I'll remove it.

opr commented on packages/checkout/components/store-notices-container/snackbar-notices.tsx:
```diff
@@ -0,0 +1,53 @@
+			} ) }
+			onRemove={ ( noticeId: string ) => {
+				notices.forEach( ( notice ) => {
+					if ( notice.explicitDismiss && notice.id === noticeId ) {
+						removeNotice( notice.id, notice.context );
+					} else {
+						removeNotice( notice.id, notice.context );
+					}
```
The same function is executed with the same args regardless of the condition? Is this is a mistake or am I missing something?

opr commented on assets/js/data/store-notices/reducers.ts:
```diff
@@ -0,0 +1,35 @@
+	action
+) => {
+	switch ( action.type ) {
+		case 'REGISTER_CONTAINER':
```
Would it be better to add these constants to a `action-types.ts` file? That's the pattern we used in other data stores.

opr commented on assets/js/data/store-notices/index.ts:
```diff
@@ -0,0 +1,36 @@
+ * External dependencies
+ */
+import { createReduxStore, register } from '@wordpress/data';
+import { controls } from '@wordpress/data-controls';
```
What are these for and why do we need them?

mikejolley replied:
I don't actually know. This was in the other stores.

opr commented on assets/js/data/store-notices/default-state.ts:
```diff
@@ -0,0 +1,10 @@
+export interface StoreNoticesState {
+	containers: Record<
+		string,
+		React.MutableRefObject< HTMLDivElement | null >
```
I don't think we should be storing the ref in here, just the context name, everything in a data store should be serializable.

Also: 
<img width="370" alt="image" src="https://user-images.githubusercontent.com/5656702/202776915-952d360a-ff3e-4c43-8262-7053b33061c8.png">
 :D

opr commented on assets/js/data/store-notices/actions.ts:
```diff
@@ -0,0 +1,17 @@
+	return {
+		type: 'REGISTER_CONTAINER',
+		containerContext,
+		ref,
```
Storing the actual ref object is not right, the data should be serializable. Can we just use the name? I think the result is the same since we only use the object's keys here https://github.com/woocommerce/woocommerce-blocks/blob/4ea76e6a3b7d4254384e8da8252df32bf4dee55d/packages/checkout/utils/create-notice.ts#L32

opr commented on assets/js/base/context/event-emit/utils.ts:
```diff
@@ -26,8 +26,16 @@ export enum responseTypes {
+	CHECKOUT = 'wc/checkout',
+	PAYMENTS = 'wc/checkout/payments',
+	EXPRESS_PAYMENTS = 'wc/checkout/express-payments',
+	CONTACT_INFO = 'wc/checkout/contact-info',
```
nitpick: should we use the full block name (minus block) for the context? so this would be 

```suggestion
	CONTACT_INFORMATION = 'wc/checkout/contact-information',
```

opr commented on assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:
```diff
@@ -62,7 +63,14 @@ export const useStoreCartCoupons = ( context = '' ): StoreCartCoupon => {
 		applyCoupon( couponCode )
 			.then( ( result ) => {
-				if ( result === true ) {
+				if (
+					result === true &&
+					__experimentalApplyCheckoutFilter(
+						'showApplyCouponNotice',
+						true,
+						{ couponCode, context }
+					)
+				) {
```
Nice, we'll need to sync this with Points and Rewards.

Should we update this deprecation notice too https://github.com/woocommerce/woocommerce-blocks/blob/trunk/packages/checkout/filter-registry/index.ts#L49 in this PR? 

Do you think we should allow `snackbarNoticeVisibility` to also be used here, but with a deprecated notice? I think it's important to deprecate functionality properly, but happy to hear your thoughts on that.

mikejolley replied:
I think that given the `__experimental` prefix we could replace this without warning. Correct me if I am wrong, but the only side effect to this filter being removed will be a notice displaying right? In which case it's not breaking, just an annoyance. We'll create a patch in points and rewards before this goes live.

opr replied:
Ok that sounds reasonable, the extension also uses `__experimentalRegisterCheckoutFilters` to add them so you're right.

opr commented on assets/js/blocks/checkout/inner-blocks/checkout-contact-information-block/block.tsx:
```diff
@@ -53,6 +58,7 @@ const Block = ( {
 
 	return (
 		<>
+			<StoreNoticesContainer context={ noticeContexts.CONTACT_INFO } />
```
Please see my comment on `assets/js/base/context/event-emit/utils.ts` about possibly renaming this.

opr reviewed (changes requested):
Nice Mike, this tested well, but I think we do need to address the refs being in the data store.

We should also deprecate the `snackbarNoticeVisibility` filter properly imo.

Please see my comments below and let me know if anything is unclear or needs further testing.

mikejolley committed:
Document forceType

mikejolley committed:
Optional props are undefined

mikejolley committed:
Remove function name

mikejolley committed:
Missing condition

mikejolley committed:
Remove context prop

mikejolley committed:
Define ACTION_TYPES

mikejolley committed:
Remove controls

mikejolley committed:
Update assets/js/base/context/event-emit/utils.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

mikejolley committed:
CONTACT_INFORMATION

mikejolley committed:
Remove ref from registerContainer

mikejolley committed:
Abstract container locating methods

mikejolley committed:
pass context correctly when displaying notices

mikejolley commented:
@opr I've worked on everything you mentioned. I'll remove the test buttons before merge, and log/patch points and rewards for the new action name.

mikejolley committed:
Remove debugging buttons

opr reviewed (approved):
Nice work, thanks Mike.

I tested it using MMQ and modifying the stock status of various items in my cart.

I am wondering if we should add some testing instructions for Globalstep/release lead for when this gets merged?

mikejolley committed:
Update filter usage - remove useMemo so filter can work inline

mikejolley commented:
@opr Compatibility patch: https://github.com/woocommerce/woocommerce-points-and-rewards/pull/491

There was a `useMemo` in our filter function which did not work when using filters inline (invalid hook usage). I've removed it — it looks like there is a cached value already so I'm not sure why useMemo was there anyway. Can you check?

mikejolley commented:
> I tested it using MMQ and modifying the stock status of various items in my cart.

@opr By modifying cart items quantities? We can also trigger a snackbar with an invalid postcode, but we should be fixing that soon I think.

opr commented:
> @opr By modifying cart items quantities? We can also trigger a snackbar with an invalid postcode, but we should be fixing that soon I think.

Yep, while on the Checkout page, I set the min quantity of an item then tried checking out. It shows the snackbar notice.
<img width="448" alt="image" src="https://user-images.githubusercontent.com/5656702/203072061-f4a11623-8c39-4398-bc91-8aa8ec6adca8.png">

Setting an incorrect postcode also shows a snackbar.

opr commented on packages/checkout/filter-registry/index.ts:
```diff
@@ -211,42 +198,40 @@ export const __experimentalApplyCheckoutFilter = < T >( {
 	/** Function that needs to return true when the filtered value is passed in order for the filter to be applied. */
 	validation?: ( value: T ) => true | Error;
 } ): T => {
-	return useMemo( () => {
```
I just tested this vs. trunk with some filters and they get called the same number of times. I think this is fine to remove.

mikejolley committed:
Refactor existing error notices from the API (#7728)

* Update API type defs

* Move create notice utils

* Replace useCheckoutNotices with new contexts

* processCheckoutResponseHeaders should check headers are defined

* Scroll to error notices only if we're not editing a field

* Error handling utils

* processErrorResponse when pushing changes

* processErrorResponse when processing checkout

* remove formatStoreApiErrorMessage

* Add todo for cart errors

* Remove unused deps

* unused imports

* Fix linting warnings

* Unused dep

* Update assets/js/types/type-defs/api-response.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Add todo

* Use generic

* remove const

* Update array types

* Phone should be in address blocks

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

mikejolley commented:
@opr I've merged the other notice PR here now - there are 2 tasks blocking this from merging because I don't want to surface too many inline notices as you populate the checkout. The worse ones are email and postcode.

opr reviewed (commented):
@mikejolley sorry to be pedantic but could we rename the directory **or** store key for the new notices store? Currently the directory names map to  the store name, prefixed with `wc/store`. Based on this folder name I, the data store should be `wc/store/store-notices`. So let's either update the key or dir name.

opr commented on assets/js/types/type-defs/api-response.ts:
```diff
@@ -0,0 +1,37 @@
+	throw new Error( 'Response not valid' );
+}
+
+export function assertResponseIsValid< T >(
```
This fails when submitting the checkout form, causing the order to be placed correctly on the server, but the client shows an error.

opr replied:
A checkout response looks like this
<img width="712" alt="image" src="https://user-images.githubusercontent.com/5656702/204620947-e4f1f2e8-a781-4ca6-b6aa-bef8bcc1a3f7.png">

mikejolley replied:
Fixed

mikejolley committed:
Update store name to wc/store/store-notices

mikejolley committed:
Fix assertResponseIsValid

mikejolley committed:
Funnel woocommerce_rest_invalid_email_address to the correct place

mikejolley committed:
woocommerce_rest_missing_email_address

mikejolley committed:
Move comments around

mikejolley committed:
Move data back into const

mikejolley committed:
Spacing

mikejolley committed:
Remove spacing

mikejolley committed:
Remove forced snack bar and styling

mikejolley committed:
Move notices within wrapper

mikejolley committed:
Remove type

mikejolley committed:
hasStoreNoticesContainer rename

mikejolley committed:
Group by status/context

mikejolley committed:
Remove global context

mikejolley committed:
Remove white space

mikejolley committed:
remove changes to simplify diff

mikejolley committed:
white space

mikejolley commented:
@opr This is rebased now, and I've added a description of all the changes I've introduced. There are quite a few files touched, but the changes are mostly minimal. Mostly typescript.

Main changes are new contexts, the merging of notice containers, and the grouping of notices by type.

opr commented on assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:
```diff
@@ -21,39 +17,21 @@ import { useStoreCart } from './use-store-cart';
  * action for adding a coupon _to_ the cart.
  * See also: https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/src/RestApi/StoreApi
- *
- * @return {StoreCartCoupon} An object exposing data and actions from/for the
```
Do you think it's worth keeping the wording here, just getting rid of the type?

mikejolley replied:
I've moved it inline with the `StoreCartCoupon` type definition.

opr commented on docs/third-party-developers/extensibility/checkout-block/available-filters.md:
```diff
@@ -212,29 +186,10 @@ __experimentalRegisterCheckoutFilters( 'automatic-coupon-extension', {
 | -------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------- |
 | ![image](https://user-images.githubusercontent.com/5656702/123768988-bc55eb80-d8c0-11eb-9262-5d629837706d.png) | ![image](https://user-images.githubusercontent.com/5656702/124126048-2c57a380-da72-11eb-9b45-b2cae0cffc37.png) |
 
-### Hide a snackbar notice containing a certain string
```
Could we document the `showApplyCouponNotice` and `showRemoveCouponNotice` filters in this PR?

mikejolley replied:
Added!

opr reviewed (commented):
Tested this and it worked according to testing instructions.

I think we'll need some design guidance because the notice looks a little weird with the icon and then the list icons, but that can be addressed in another PR, functionally this is OK.

<img width="593" alt="image" src="https://user-images.githubusercontent.com/5656702/207927021-219b3141-8d8f-45cb-95d4-4c7da5be8420.png">

Added a couple of inline comments that are non-blocking, and maybe the documentation for the filter could be added in another PR. Approving.

opr reviewed (approved):
Didn't hit approve in the last PR when posting the review 😖

mikejolley committed:
Move comment to typescript

mikejolley commented:
@opr that list styling shouldn't be there, so Ill remove that at least and then await design input for other styling changes.

mikejolley committed:
List style

mikejolley committed:
showApplyCouponNotice docs

mikejolley committed:
See if scrollIntoView exists

mikejolley committed:
fix notice tests
